File: app/ui/src/types/documents.ts
export type DocumentJobStatus =
  | "pending"
  | "running"
  | "completed"
  | "failed"
  | "cancelled";

export interface DocumentListItem {
  id: number;
  guid: string;
  name: string;
  aircraft_model_name: string | null;
  latest_job_status: DocumentJobStatus | null;
  serial_ranges: SerialRangeResponse[];
  created_at: string;
}

export interface DocumentListResponse {
  documents: DocumentListItem[];
  total: number;
}

export interface DocumentVersionDetail {
  guid: string;
  file_name: string;
  name: string;
  content_type: string | null;
  file_size: number | null;
  blob_path: string | null;
}

export type SerialRangeType = "single" | "range" | "and_subs";

export interface SerialRangeResponse {
  id: number;
  range_type: SerialRangeType;
  serial_start: string;
  serial_end: string | null;
}

export interface DocumentDetailResponse {
  guid: string;
  name: string;
  aircraft_model_id: number | null;
  aircraft_model_name: string | null;
  document_category_id: number | null;
  document_category_name: string | null;
  document_type_id: number | null;
  document_type_name: string | null;
  latest_version: DocumentVersionDetail | null;
  serial_ranges: SerialRangeResponse[];
}

export interface SerialRangeInput {
  range_type: SerialRangeType;
  serial_start: string;
  serial_end?: string;
}

export interface DocumentUpdateRequest {
  name?: string;
  document_category_id?: number;
  document_type_id?: number;
  serial_ranges?: SerialRangeInput[];
}

export interface DocumentDownloadUrlResponse {
  download_url: string;
  version_name: string;
  file_name: string;
  file_size: number | null;
  content_type: string | null;
}

File: app/api/schemas/documents.py
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel


class SerialRangeInput(BaseModel):
    range_type: Literal["single", "range", "and_subs"]
    serial_start: str
    serial_end: Optional[str] = None


class SerialRangeResponse(BaseModel):
    id: int
    range_type: Literal["single", "range", "and_subs"]
    serial_start: str
    serial_end: Optional[str] = None

    class Config:
        from_attributes = True


class DocumentListItem(BaseModel):
    id: int
    guid: str
    name: str
    aircraft_model_name: Optional[str] = None
    latest_job_status: Optional[str] = None
    serial_ranges: list[SerialRangeResponse] = []
    created_at: datetime

    class Config:
        from_attributes = True


class DocumentListResponse(BaseModel):
    documents: list[DocumentListItem]
    total: int


class DocumentVersionDetail(BaseModel):
    guid: str
    name: str
    file_name: str
    content_type: Optional[str] = None
    file_size: Optional[int] = None
    blob_path: Optional[str] = None

    class Config:
        from_attributes = True


class DocumentDetailResponse(BaseModel):
    guid: str
    name: str
    aircraft_model_id: Optional[int] = None
    aircraft_model_name: Optional[str] = None
    document_category_id: Optional[int] = None
    document_category_name: Optional[str] = None
    document_type_id: Optional[int] = None
    document_type_name: Optional[str] = None
    latest_version: Optional[DocumentVersionDetail] = None
    serial_ranges: list[SerialRangeResponse] = []

    class Config:
        from_attributes = True


class DocumentUpdateRequest(BaseModel):
    name: Optional[str] = None
    document_category_id: Optional[int] = None
    document_type_id: Optional[int] = None
    serial_ranges: Optional[list[SerialRangeInput]] = None


class DocumentDownloadUrlResponse(BaseModel):
    download_url: str
    file_name: str
    file_size: Optional[int] = None
    content_type: Optional[str] = None
